components/capi/src/provider.rs

```diff
@@ -0,0 +1,80 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_provider::erased::ErasedDataProvider as ICUErasedProvider;
+use icu_provider_fs::FsDataProvider;
+use std::{mem, ptr, slice, str};
+
+#[repr(C)]
+/// Safety: This should only
+pub struct ErasedDataProvider {
+    /// Dummy fields to ensure this is the size of a trait object pointer
+    /// Can be improved once the Metadata API stabilizes
+    _field1: usize,
+    _field2: usize,
+}
+
+impl ErasedDataProvider {
+    /// This is unsafe because zeroed() can be passed to other functions
+    /// and cause UB
+    ///
+    /// This is necessary for returning uninitialized values to C
+    pub unsafe fn zeroed() -> Self {
+        ErasedDataProvider {
+            _field1: 0,
+            _field2: 0,
+        }
+    }
+
+    pub fn from_boxed(x: Box<dyn ICUErasedProvider<'static>>) -> Self {
+        unsafe {
+            // if the layout changes this will error
+            mem::transmute(x)
+        }
+    }
+
+    pub fn into_boxed(self) -> Box<dyn ICUErasedProvider<'static>> {
+        unsafe { mem::transmute(self) }
+    }
+
+    pub fn as_dyn_ref(&mut self) -> &mut dyn ICUErasedProvider<'static> {
+        unsafe {
+            // &mut dyn Trait and Box<dyn Trait> have the same layout
+            let borrowed_erased: ErasedDataProvider = ptr::read(self);
+            mem::transmute(borrowed_erased)
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn erased_data_provider_destroy(d: ErasedDataProvider) {
+    let _ = d.into_boxed();
+}
+
+#[repr(C)]
+pub struct CreateDataProviderResult {
+    provider: ErasedDataProvider,
+    success: bool,
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fs_data_provider_create(
```

Comment:
If I understand your code correctly... I'm pleasantly shocked that this works!  This is great!  I think what it means is that this `fs_data_provider_create` sets up the reader from the filesystem without including any data struct deserializers, and the deserializers live in the constructor functions like `plural_rules_create()`.  This is sweet, because it means we don't need to add a whole bunch of things like `DataProvider_PluralRulesV1` as we were discussing earlier.

It also means that Serde is always enabled, potentially increasing code size... but that's a problem we can solve later.  If we decide to make the C API always use ErasedDataProvider, we can simply make a version of ErasedDataProvider without Serde if needed.

Reply:
> This is sweet, because it means we don't need to add a whole bunch of things like DataProvider_PluralRulesV1 as we were discussing earlier

yeah, that's why I mentioned I was going to use Erased for a first cut.

We now _can_ add specific APIs for `DataProvider_PRv1`, but we don't _have_ to (I think)

Reply:
I think we should stick with ErasedDataProvider.  According to my benchmarks, the performance hit is small (5-10% on a load operation), so it is likely to be absorbed (no measurable impact on a large-scale system).

https://github.com/unicode-org/icu4x/pull/431#issuecomment-750780808

Reply:
Good to hear! In that case we can stick to Erased for now!